Move FlushProcessWriteBuffers to minipal

radekdoulik opened the pull request:
Implements https://github.com/dotnet/runtime/issues/117977

Also move the initialization of it and get of page size

radekdoulik committed:
Move FlushProcessWriteBuffers to minipal

Implements https://github.com/dotnet/runtime/issues/117977

Also move the initialization of it and get of page size

radekdoulik committed:
Merge branch 'main' into coreclr-move-FlushProcessWriteBuffers-to-minipal

radekdoulik committed:
Fix build

am11 commented on src/native/minipal/memory.h:
```diff
@@ -0,0 +1,37 @@
+{
+#endif // __cplusplus
+
+    bool minipal_initialize_flush_process_write_buffers();
+    void minipal_shutdown_flush_process_write_buffers();
+    void minipal_flush_process_write_buffers();
```
nit: C style prototypes:
```suggestion
    bool minipal_initialize_flush_process_write_buffers(void);
    void minipal_shutdown_flush_process_write_buffers(void);
    void minipal_flush_process_write_buffers(void);
```

am11 commented on src/native/minipal/memory.c:
```diff
@@ -0,0 +1,235 @@
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdio.h>
```
```suggestion
#include <stdint.h>
#include <stdio.h>
```

jkotas commented on src/native/minipal/memory.c:
```diff
@@ -0,0 +1,235 @@
+{
+    int pageSize = sysconf( _SC_PAGE_SIZE );
+
+    g_pageSizeUnixInl = (size_t)((pageSize > 0) ? pageSize : 0x1000);
```
`g_pageSizeUnixInl` has nothing to do flush process write buffers. It is odd to initialize it in `minipal_initialize_flush_process_write_buffers`.

We may want to avoid touching page size in this PR, and just

radekdoulik replied:
this was just convenient place to use, which is called during initialization. we could introduce minipal initialize to have better place for it. the idea was to deduplicate page size as well. otoh it is small part so I will leave it out then.

radekdoulik committed:
Fix linux build

radekdoulik committed:
Update definitions

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

am11 commented on src/native/minipal/memory.c:
```diff
@@ -0,0 +1,236 @@
+
+size_t g_pageSizeUnixInl = 0;
+
+bool minipal_initialize_flush_process_write_buffers()
```
```suggestion
bool minipal_initialize_flush_process_write_buffers(void)
```

am11 commented on src/native/minipal/memory.c:
```diff
@@ -0,0 +1,236 @@
+    return true;
+}
+
+void minipal_shutdown_flush_process_write_buffers()
```
```suggestion
void minipal_shutdown_flush_process_write_buffers(void)
```

am11 commented on src/native/minipal/memory.c:
```diff
@@ -0,0 +1,236 @@
+}
+
+// Flush write buffers of processors that are executing threads of the current process
+void minipal_flush_process_write_buffers()
```
```suggestion
void minipal_flush_process_write_buffers(void)
```

am11 commented on src/native/minipal/memory.c:
```diff
@@ -0,0 +1,236 @@
+#endif // !TARGET_WASM
+}
+#else // !TARGET_WINDOWS
+void minipal_flush_process_write_buffers()
```
```suggestion
void minipal_flush_process_write_buffers(void)
```

am11 commented on src/native/minipal/memory.h:
```diff
@@ -0,0 +1,37 @@
+#endif
+
+#ifndef DACCESS_COMPILE
+    __forceinline size_t minipal_get_page_size()
```
```suggestion
    __forceinline size_t minipal_get_page_size(void)
```

radekdoulik committed:
Feedback

Do not care about cleanup

radekdoulik committed:
Feedback

Do not move page size to minipal

radekdoulik committed:
Feedback

Rename the files

radekdoulik committed:
Feedback

Changes which were missed in earlier commits

radekdoulik committed:
Merge branch 'main' into coreclr-move-FlushProcessWriteBuffers-to-minipal

jkotas commented on src/coreclr/pal/src/init/pal.cpp:
```diff
@@ -580,7 +581,7 @@ Initialize(
         {
             // Initialize before first thread is created for faster load on Linux
-            if (!InitializeFlushProcessWriteBuffers())
+            if (!minipal_initialize_flush_process_write_buffers())
```
I think `minipal_initialize_flush_process_write_buffers` needs to be called from Mono initialization, standalone GC initialization and/or NativeAOT initialization as well.

jkotas replied:
Sounds reasonable to me.

jkotas commented on src/native/minipal/flushprocesswritebuffers.c:
```diff
@@ -0,0 +1,226 @@
+{
+    int pageSize = sysconf( _SC_PAGE_SIZE );
+
+    s_pageSize = (size_t)((pageSize > 0) ? pageSize : 0x1000);
```
We need the page size only on the fallback path that uses g_helperPage. Initialization should be moved there.

radekdoulik committed:
Fix mac build

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

radekdoulik committed:
Feedback

Remove PalFlushProcessWriteBuffers wrapper

radekdoulik committed:
Feedback

Remove parts around g_helperPage on mac

radekdoulik committed:
Feedback

Rename to memory_barrier_process_wide

radekdoulik committed:
Merge branch 'main' into coreclr-move-FlushProcessWriteBuffers-to-minipal

radekdoulik committed:
Fix build

radekdoulik committed:
Fix windows build

radekdoulik committed:
More specific include

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

radekdoulik committed:
Fix windows build

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

radekdoulik committed:
Merge branch 'main' into coreclr-move-FlushProcessWriteBuffers-to-minipal

jkotas commented on src/coreclr/nativeaot/Runtime/windows/PalMinWin.cpp:
```diff
@@ -1039,7 +1039,7 @@ UInt32_BOOL PalCloseHandle(HANDLE arg1)
 }
 
-void PalFlushProcessWriteBuffers()
+void minipal_memory_barrier_process_wide()
```
Should this be deleted instead? I would expect the implementation to come from minipal.

radekdoulik committed:
Feedback

Remove windows wrapper too

am11 commented on src/native/minipal/memorybarrierprocesswide.c:
```diff
@@ -0,0 +1,232 @@
+}
+#else // !HOST_WINDOWS
+
+#include <processthreadsapi.h>
```
Looks like we need the full thing (or figure out the per-arch define..):
```suggestion
#include <windows.h>
```

jkotas commented on src/native/minipal/memorybarrierprocesswide.c:
```diff
@@ -0,0 +1,232 @@
+#endif
+#endif
+
+static bool CanFlushUsingMembarrier(void)
+{
+#if defined(__linux__) || HAVE_SYS_MEMBARRIER_H
```
```suggestion
#if HAVE_SYS_MEMBARRIER_H
static bool CanFlushUsingMembarrier(void)
{
```

jkotas commented on src/native/minipal/memorybarrierprocesswide.c:
```diff
@@ -0,0 +1,232 @@
+    return false;
+}
+
+//
+// Tracks if the OS supports FlushProcessWriteBuffers using membarrier
+//
+static int s_flushUsingMemBarrier = 0;
+
+#ifndef TARGET_APPLE
```
```suggestion
#if HAVE_SYS_MEMBARRIER_H
//
// Tracks if the OS supports membarrier syscall
//
static int s_flushUsingMemBarrier = 0;
#endif // HAVE_SYS_MEMBARRIER_H

#if !defined(HOST_APPLE) && !defined(HOST_WASM)
```
s_flushUsingMemBarrier can have more precise ifdef.

jkotas commented on src/native/minipal/memorybarrierprocesswide.c:
```diff
@@ -0,0 +1,232 @@
+#define membarrier(...) syscall(__NR_membarrier, __VA_ARGS__)
+#elif HAVE_SYS_MEMBARRIER_H
+#include <sys/membarrier.h>
+#ifdef TARGET_BROWSER
+#define membarrier(cmd, flags, cpu_id) 0 // browser/wasm is currently single threaded
+#endif
```
```suggestion
```
Is this actually needed?

jkotas commented on src/native/minipal/memorybarrierprocesswide.c:
```diff
@@ -0,0 +1,232 @@
+
+bool minipal_initialize_memory_barrier_process_wide(void)
+{
+#ifndef TARGET_WASM
```
```suggestion
#ifdef HOST_WASM
    // browser/wasm is currently single threaded
#elif defined(HOST_APPLE)
...

#else
#if HAVE_SYS_MEMBARRIER_H
    if (CanFlushUsingMembarrier())
    {
        s_flushUsingMemBarrier = true;
    }
    else
#else
    {
        // Fallback implementation
    }
#endif
```
I think conditions with positive polarity would make the code more readable.

Same structure for minipal_memory_barrier_process_wide

radekdoulik committed:
Use more generic include again

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

radekdoulik committed:
Apply suggestions from code review

Refactor ifdefs

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

radekdoulik committed:
Merge branch 'main' into coreclr-move-FlushProcessWriteBuffers-to-minipal

radekdoulik committed:
Fix build

radekdoulik committed:
Complete the suggestion from feedback

radekdoulik committed:
Feedback

Call `minipal_initialize_memory_barrier_process_wide` from unix
`GCToOSInterface::Initialize`. Let it return true if it was previously
successfully initialized.

radekdoulik committed:
Fix windows build

jkotas commented on src/native/minipal/memorybarrierprocesswide.c:
```diff
@@ -0,0 +1,238 @@
+//
+// Tracks if the OS supports membarrier syscall
+//
+static int s_flushUsingMemBarrier = 0;
```
```suggestion
static bool s_flushUsingMemBarrier = false;
```

jkotas commented on src/native/minipal/memorybarrierprocesswide.c:
```diff
@@ -0,0 +1,238 @@
+#endif // HAVE_SYS_MEMBARRIER_H
+
+#if !defined(HOST_APPLE) && !defined(HOST_WASM)
+// Helper memory page used by the FlushProcessWriteBuffers
```
```suggestion
// Helper memory page used by the fallback path
```

jkotas commented on src/native/minipal/memorybarrierprocesswide.c:
```diff
@@ -0,0 +1,238 @@
+static pthread_mutex_t g_flushProcessWriteBuffersMutex;
+
+static size_t s_pageSize = 0;
+#endif // !TARGET_APPLE
```
```suggestion
#endif // !TARGET_APPLE && !HOST_WASM
```

jkotas commented on src/native/minipal/memorybarrierprocesswide.c:
```diff
@@ -0,0 +1,238 @@
+// Helper memory page used by the FlushProcessWriteBuffers
+static uint8_t* g_helperPage = 0;
+
+// Mutex to make the FlushProcessWriteBuffersMutex thread safe
```
```suggestion
// Mutex to make the fallback path thread safe
```

jkotas commented on src/native/minipal/memorybarrierprocesswide.c:
```diff
@@ -0,0 +1,238 @@
+            return false;
+        }
+
+        // Verify that the s_helperPage is really aligned to the g_SystemInfo.dwPageSize
```
```suggestion
        // Verify that the s_helperPage is really aligned to the s_pageSize
```

jkotas commented on src/native/minipal/memorybarrierprocesswide.c:
```diff
@@ -0,0 +1,238 @@
+// Flush write buffers of processors that are executing threads of the current process
+void minipal_memory_barrier_process_wide(void)
+{
+#ifndef TARGET_WASM
```
```suggestion
    assert(s_initializedMemoryBarrierSuccessfullyInitialized);

#ifndef TARGET_WASM
```

jkotas replied:
I would follow the same ifdef structure here as in initialize method:

```
#ifdef HOST_WASM
    // browser/wasm is currently single threaded
#elif defined(HOST_APPLE)
    ...
#else
...
```

jkotas commented on src/native/minipal/memorybarrierprocesswide.c:
```diff
@@ -0,0 +1,238 @@
+    }
+#endif // !HOST_WASM && !HOST_APPLE
+
+    return s_initializedMemoryBarrierSuccessfullyInitialized = true;
```
```suggestion
    s_initializedMemoryBarrierSuccessfullyInitialized = true;
    return true;
```
`return s_initializedMemoryBarrierSuccessfullyInitialized = true;` is likely to make static analyzers complain about potential bug (`=` vs. `==`).

BrzVlad commented:
@radekdoulik Note that I checked locally and this PR seems to fix https://github.com/dotnet/runtime/issues/115480.

radekdoulik committed:
Apply suggestions from code review

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

radekdoulik committed:
Merge branch 'main' into coreclr-move-FlushProcessWriteBuffers-to-minipal

radekdoulik committed:
Feedback

Refactor minipal_memory_barrier_process_wide

radekdoulik committed:
Fix build

jkotas committed:
Apply suggestions from code review

jkotas commented on src/native/minipal/memorybarrierprocesswide.c:
```diff
@@ -0,0 +1,240 @@
+        // Fallback implementation
+        assert(g_helperPage == NULL);
+
+        int pageSize = sysconf( _SC_PAGE_SIZE );
```
Ok

radekdoulik committed:
Merge branch 'main' into coreclr-move-FlushProcessWriteBuffers-to-minipal

radekdoulik committed:
Merge branch 'main' into coreclr-move-FlushProcessWriteBuffers-to-minipal
